components/movementsensor/wheeledodometry/wheeled_odometry.go

```diff
@@ -0,0 +1,302 @@
+	}
+	o.orientation.Yaw = 0
+	o.math(context.Background())
+
+	return o, nil
+}
+
+func (o *odometry) AngularVelocity(ctx context.Context, extra map[string]interface{}) (spatialmath.AngularVelocity, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.angularVelocity, nil
+}
+
+func (o *odometry) LinearAcceleration(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.acceleration, nil
+}
+
+func (o *odometry) Orientation(ctx context.Context, extra map[string]interface{}) (spatialmath.Orientation, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	ov := &spatialmath.OrientationVector{Theta: o.orientation.Yaw, OX: 0, OY: 0, OZ: 1}
+	return ov, nil
+}
+
+func (o *odometry) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.orientation.Yaw, nil
+}
+
+func (o *odometry) LinearVelocity(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.linearVelocity, nil
+}
+
+func (o *odometry) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return geo.NewPoint(o.position.X, o.position.Y), o.position.Z, nil
+}
+
+func (o *odometry) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	return movementsensor.Readings(ctx, o, extra)
+}
+
+func (o *odometry) Accuracy(ctx context.Context, extra map[string]interface{}) (map[string]float32, error) {
+	return map[string]float32{}, movementsensor.ErrMethodUnimplementedAccuracy
+}
+
+func (o *odometry) Properties(ctx context.Context, extra map[string]interface{}) (*movementsensor.Properties, error) {
+	return &movementsensor.Properties{
+		LinearVelocitySupported:     true,
+		AngularVelocitySupported:    true,
+		OrientationSupported:        true,
+		PositionSupported:           true,
+		LinearAccelerationSupported: false,
+	}, nil
+}
+
+func (o *odometry) Close(ctx context.Context) error {
+	o.cancelFunc()
+	o.activeBackgroundWorkers.Wait()
+	return nil
+}
+
+func (o *odometry) math(ctx context.Context) {
+	o.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer utils.UncheckedErrorFunc(func() error {
+			o.mu.Lock()
+			defer o.mu.Unlock()
+			return errors.New("math failed")
+		})
+		defer o.activeBackgroundWorkers.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			leftMove, err := o.motors[0].left.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+			}
+			rightMove, err := o.motors[0].right.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+			}
+			if leftMove || rightMove {
+				// sleep for the desired time interval
+				time.Sleep(time.Duration(o.timeInterval) * time.Millisecond)
+
+				// use GetInParallel to ensure the left and right motors are polled at the same time
```

Comment:
is there part of this that you would like added to the comment? I didn't think we would add specific information about modular components to RDK since they are separate but I definitely can if you think that's important to point out?

Reply:
The first time I read this, I said to myself "it takes more time to create a coroutine than it does to grab the info from another component within the RDK. This probably isn't harmful, but it's at best wishful thinking." It wasn't until later that I realized that, for modular (non-RDK) components, this is actually important. Perhaps that will be more obvious in a few months when modular components are more common.

I dunno; leave it as-is if you prefer. :shrug: